src/csharp/Grpc.Core.Tests/Internal/AsyncCallTest.cs

```diff
@@ -472,6 +472,145 @@ public void ServerStreaming_MoreResponses_Success()
+            // are released correctly no matter the order of the calls, and that when the call is cancelled
+            // or there is an error (such as server unavailable) that MoveNext throws the correct exception.
+
+            // This test is testing the order:
+            // OnReceivedResponseHeaders
+            // Cancel (which could be because of a Dispose or explicit cancel)
+            // OnReceivedStatusOnClient
+            // MoveNext - fails with Exception
+
+            asyncCall.StartServerStreamingCall("request1");
+            var responseStream = new ClientResponseStream<string, string>(asyncCall);
+
+            // Check that the client has a reference to exactly one call
+            Assert.AreEqual(1, channel.GetCallReferenceCount());
+            // Check that the call is not yet disposed
+            Assert.IsFalse(fakeCall.IsDisposed);
+
+            // Cancel the call. This will trigger ReceivedStatusOnClientCallback to be called with a Cancelled status.
+            asyncCall.Cancel();
+            fakeCall.ReceivedStatusOnClientCallback.OnReceivedStatusOnClient(true, new ClientSideStatus(Status.DefaultCancelled, new Metadata()));
+
+            // OnReceivedResponseHeaders with have success == false as call is cancelled
+            fakeCall.ReceivedResponseHeadersCallback.OnReceivedResponseHeaders(false, new Metadata());
+            Assert.AreEqual(0, asyncCall.ResponseHeadersAsync.Result.Count);
+
+            // read some data - this should fail as the call has been cancelled
+            var ex = Assert.ThrowsAsync<RpcException>(async () => await responseStream.MoveNext());
+            Assert.AreEqual(Status.DefaultCancelled, ex.Status);
+
+            // Check that the resources have been released and that the client no longer references the call
+            Assert.AreEqual(0, channel.GetCallReferenceCount());
+            Assert.IsTrue(fakeCall.IsDisposed);
+        }
+
+        [Test]
+        public void ServerStreaming_CancelCall3()
+        {
+            // Test client cancelling the call before the reading the server stream to the end.
+            // Test for resource leak https://github.com/grpc/grpc/issues/8451
+
+            // The callbacks OnReceivedResponseHeaders, OnReceivedStatusOnClient and OnReceivedMessage may
+            // occur in any order together with calls to MoveNext. We need to test that the resources
+            // are released correctly no matter the order of the calls, and that when the call is cancelled
+            // or there is an error (such as server unavailable) that MoveNext throws the correct exception.
+
+            // This test is testing the order:
+            // OnReceivedResponseHeaders
+            // Cancel (which could be because of a Dispose or explicit cancel)
+            // MoveNext - OK as cancel status not yet picked up in callback
+            // OnReceivedMessage
+            // OnReceivedStatusOnClient - delayed callback from the Cancel
```

Comment:
In all the test case, you always have OnReceivedMessage be delivered before OnReceivedStatusOnClient, but that's not something that's generally true. Status can be delivered while there is an inflight read operation,
so MoveNext -> OnReceivedStatusOnClient->OnReceivedMessage is also a legal sequence. If that happens, the native resources should be kept alive during OnReceivedStatusOnClient and they should be released during the final OnReceivedMessage.
Can you please add a test for that?

Reply:
I've added further tests